### DISCUSSION TITLE
is it possible to call functions from the frontend directly to the +page.server with on:click?

### DISCUSSION THREAD

**Discussion by ctwhome** [sveltejs/kit#7796] [Q&A] [answered]
I am looking for a way to call server-side functions, not necessarily when loading the page, but when a button is clicked in the front end. Right now I am using cloud functions to execute these kinds of functions, but I wonder if I can just use the server from sveltekit?

**Comment by apexfactory** [comment]
You can import a function that you've defined in your back end code, but I'm not sure that's exactly what you mean.

**Reply by machak** [reply]
I think he mains something like https://en.wikipedia.org/wiki/Java_remote_method_invocation

**Comment by infovore** [comment, marked as answer]
Sure. Make a server-side endpoint and `POST` to it (or `GET` if you must...); or use a page endpoint and a an [action](https://kit.svelte.dev/docs/form-actions), just like you would calling... any other server side function?

**Comment by ctwhome** [comment]
Thank you for all the answers. I went further to study how action forms and endpoints really work. I just couldn't believe that the endpoints would become a cloud function (in Vercel)! Just wow.

I had to wrap my head around these concepts. So the answer to my question is yes. You can call with the click of a button any function that lives in the server (+server.ts, +page.ts, +page.server.ts) but with a slightly different method: either using form actions or calling the endpoint as @infovore said. thanks so much. I'll try to add a working example on the template I am building https://github.com/ctwhome/top-sveltekit

**Reply by david-plugge** [reply]
Nice project, happy to help you when it comes to integrating supabase :)

**Reply by infovore** [reply]
Sending HTTP messages from clients to server - GETs via links, POSTs via buttons and forms - is... just how the web works, and has worked, for decades? And in a full-stack framework like SvelteKit, you're responsible for both client and server.

The automatic Vercel integration comes entirely down to the _adapter_ that's used when it is deployed. `adapter-node` is bascially equivalent what you have in development mode: your server functions running on a node/express server, and that server also serving front-end content as static files when appropriate. When you deploy to Vercel/Netlify/Cloudflare, `adapter-auto` (or the relevant adapter) will instead deploy your static/prerendered content as static files, but extract your server endpoints (whether they're 'raw' `+server.ts` files or page endoints like `+page.server.ts`) to cloud-function endpoints in the appropriate style. Broadly speaking, It Just Works, but it's useful to know the limitations of your individual provider as not every function that is Just Fine in `adapter-node` ports to cloud providers instantly. For instance, there's a 10s limit on function execution at Netlify, and also a limit on the payload you can return. But the adapters approach is great; as designed, you write a tradiational client/server application, and then the translation of that to a deployment environment is, hopefully, not your problem.

**Reply by ctwhome** [reply]
this is gold @infovore, thanks so much for taking the time to answer in detail. So I was deploying my functions to firebase functions...oh my, imagine my face when understanding that all of this come together from just SvelteKit

**Reply by infovore** [reply]
well, depending on how you host it! I'm not sure if you're using Firebase by preference or by necessity, but the unofficial Firebase adapter is still highly beta-ish, and I think is waiting for SK to come out of beta before it settles down. There's a perfectly acceptable App Engine adapter that basically wraps `adapter-node` with an `app.yaml` file, if you have an incentive to use GCP.

**Reply by ctwhome** [reply]
I used it by necessity, to keep a webhook always listening. I don't think I would use GCP directly, since Firebase creates a nice wrapper around it for what I needed (hosting-auth-NoSQL-storage-functions). I have used Firebase in the past extensively but I can carry on with endpoints being converted into cloud functions. But I am now moving the stack of two projects to Supabse, (The overengineering I was doing with Firestore was killing me, and also I had severe limitations). 
I am curious what is your tech stack lately!

**Reply by infovore** [reply]
Tech stack varies depending on client and project, but: for $reasons, my main project right now uses Firebase for storage, auth, and firestore, and deploys to App Engine using [this adapter](https://github.com/HalfdanJ/svelte-adapter-appengine). Other recent projects have either been entirely static, deploying to services like Netlify, or in one significant case, `adapter-node` running on Digital Ocean App Platform, but with a significant number (thousands) of prerendered pages.
